proposals/TP1-xds-transport-next.md

```diff
@@ -0,0 +1,1028 @@
+
+We distinguish between two forms of `xdstp://` URIs:
+
+* Uniform Resource Locators (URLs) instructing a client on how to locate a given
+  xDS resource. URLs may include processing directives.
+
+* Uniform Resource Names (URNs) referring to a specific xDS resource. URNs do not
+  include processing directives. `xdstp://` URNs provide self-contained cache keys.
+
+#### Proto3 expression
+
+`xdstp://` URIs are canonically encoded in `proto3` messages. The structure
+for this encoding is provided at https://github.com/cncf/xds/tree/main/xds/core/v3:
+
+* [URL](https://github.com/cncf/xds/blob/main/xds/core/v3/resource_locator.proto) definition
+* [URN](https://github.com/cncf/xds/blob/main/xds/core/v3/resource_name.proto) definition
+
+#### Normalization
+
+Two `xdstp://` URNs are considered equivalent if they match component-wise
+modulo context parameter ordering.
+
+## Context parameters
+
+Context parameters in URNs presented by the client to the server will be dervied
+by the following process. Using an example of a URL
+`xdstp://some-authority/some.type/foo?xds.bar=baz`:
+
+1. Static `Node`-derived context parameters are first added. These are prefixed with
+   `xds.node.`. The set of `Node`-derived context parameters is specified in the
+   bootstrap on a per-resource type basis. The key reflects the `Node` proto3
+   field structure, e.g. `xds.node.locality.sub_zone=some_sub_zone`,
+   `xds.node.metadata.bar="a"`, `xds.node.user_agent_version=1.2.3`.
+
+2. Dynamic context parameters are added next. These are free form and at the
+   discretion of the client. They can provide client-specific fields such as
+   `xds.shard_id=1234` and potentially change through the lifetime of an xDS
+   stream.
+
+3. Context parameters from the URL are added next, in the above example
+   `xds.bar=baz`.
+
+4. [Client feature
+   capability](https://www.envoyproxy.io/docs/envoy/latest/api/client_features)
+   context parameters are then added, e.g. clusters at the client might have
+   `{xds.client_feature.lb.least_loaded: true}`. Client feature capabilities
+   will be required to be declared with respect to a specific resource type URL.
+   All these attributes will be `xds.client_feature.` prefixed.
+
+5. Finally per-resource type well-known attributes are added, e.g. an on-demand
+   listener load might have `{xds.resource.vip: 96.54.3.1}`. These attributes
+   are generated by code, e.g. an on-demand LDS implementation will first detect
+   the VIP, halt the request, ask for the `xds.resource.vip` from the control
+   plane, and then resume the request after discovery is completed. All these
+   attributes will be `xdstp.resource.` prefixed.
+
+An example computed URN following the above example is
+`xdstp://some-authority/some.type/foo?xds.node.metadata.foo=bar&xds.shard_id=1234&xds.bar=baz&xds.client_feature.lb.least_loaded=false&xds.resource.vip=96.54.3.1`.
```

Comment:
This will probably be massive for some cases - well over 2048 bytes which seems to be the limit for some libraries. But maybe that won't be an issue here since we are not HTTP. I am not familiar with enough URI parsing libraries to know if any of them will complain or if its just internet explorer which is irrelevant here.

Reply:
If names grow this large, we can encounter wire performance problems. We've spent a bit of time thinking on how to optimize this on the transport, e.g. factor out common context parameters. The thing to keep in mind is that URIs are a logical view of naming as well as concrete syntax; we can always optimize internally and on the wire (with protobuf) to avoid explicit string representation when things become unwieldy, but ideally applications limit the scope and size of context parameters.

Reply:
https://docs.google.com/document/d/1zZav-IYxMO0mP2A7y5XHBa9v0eXyirw1CxrLLSUqWR8/edit#heading=h.xbjx2ihsbbaa covers the HTTP transport. I have not included this yet, since I'm trying to keep the Markdown to a fairly limited implementable-in-the-near-term scope. We can port that stuff across when we start work on the HTTP support.